obspy/io/sac/arrayio.py

```diff
@@ -0,0 +1,637 @@
+
+    if is_file_name:
+        f.close()
+
+
+
+def write_sac_ascii(dest, hf, hi, hs, data=None):
+    """
+    Write the header and (optionally) data arrays to a SAC ASCII file.
+
+    Parameters
+    ----------
+    dest : str or File-like object
+        Full path or File-like object from SAC ASCII file on disk.
+    hf, hi, hs : numpy.ndarray
+        The float, integer, and string header arrays.
+    data : numpy.ndarray, optional
+        float32 data array.  If omitted or None, it is assumed that the user
+        intends to overwrite/modify only the header arrays of an existing file.
+        Equivalent to "writehdr". If data is None, better make sure the header
+        you're writing matches any data already in the file.
+
+    """
+    # TODO: fix prodigious use of file open/close for "with" statements.
+
+    if data is None:
+        # file exists, just modify it (don't start from scratch)
+        fmode = 'r+'
+    else:
+        # start from scratch
+        fmode = 'w+'
+
+    try:
+        f = open(dest, fmode)
+        is_file_name = True
+    except IOError:
+        raise SacIOError("Cannot open file: " + dest)
+    except TypeError:
+        f = dest
+        is_file_name = False
+
+    if data is None and (f.mode != 'r+'):
+        msg = "File mode must be 'r+' for data=None."
+        raise ValueError(msg)
+
+    try:
+        np.savetxt(f, np.reshape(hf, (14, 5)),
+                   fmt=native_str("%#15.7g%#15.7g%#15.7g%#15.7g%#15.7g"))
+        np.savetxt(f, np.reshape(hi, (8, 5)),
+                   fmt=native_str("%10d%10d%10d%10d%10d"))
+        for i in range(0, 24, 3):
+            f.write(hs[i:i + 3].data)
+            f.write(b'\n')
+    except:
+        if is_file_name:
+            f.close()
+        raise SacIOError("Cannot write header values: " + f.name)
+
+    if data is not None:
+        npts = hi[9]
+        if npts in (HD.INULL, 0):
+            if is_file_name:
+                f.close()
+            return
+        try:
+            rows = npts // 5
+            np.savetxt(f, np.reshape(data[0:5 * rows], (rows, 5)),
+                       fmt=native_str("%#15.7g%#15.7g%#15.7g%#15.7g%#15.7g"))
+            np.savetxt(f, data[5 * rows:], delimiter=b'\t')
+        except:
+            f.close()
+            raise SacIOError("Cannot write trace values: " + f.name)
+
+    if is_file_name:
+        f.close()
+
+
+# ---------------------- HEADER ARRAY / DICTIONARY CONVERTERS -----------------
+# TODO: this functionality is basically the same as the getters and setters in
+#    sac.sactrace. find a way to avoid duplication?
+# TODO: put these in sac.util?
+def header_arrays_to_dict(hf, hi, hs, nulls=False):
+    """
+    Returns
+    -------
+    dict
+        The correctly-ordered SAC header values, as a dictionary.
+    nulls : bool
+        If True, return all values including nulls.
+
+    """
+    if nulls:
+        items = [(key, val) for (key, val) in zip(HD.FLOATHDRS, hf)] + \
```

Comment:
Agreed, but list(zip(..)) obfuscates the difference between this block and the next one, for which a list comprehension seems necessary.